### PR TITLE
Add report recipe OOI

### PR DESCRIPTION
**Pull request by Rieven**
### Changes

Create e new report recipe OOI which includes:

- A set of input OOIs or a query which can dynamically fetch the necessary input to make a report
- report types to know which reports to make.
- Cron expression, this is for the scheduler to know when to schedule the recipe to make new reports.
- A report has a reference to a report recipe

### Issue link
https://github.com/minvws/nl-kat-coordination/issues/3088

Closes: Does not close any yet as this is a part of a larger part.

### QA notes

_Please add some information for QA on how to test the newly created code._

---

### Code Checklist

- [x] All the commits in this PR are properly PGP-signed and verified.
- [x] This PR only contains functionality relevant to the issue.
- [ ] I have written unit tests for the changes or fixes I made.
- [x] I have checked the documentation and made changes where necessary.
- [x] I have performed a self-review of my code and refactored it to the best of my abilities.

- [x] Tickets have been created for newly discovered issues.
- [x] For any non-trivial functionality, I have added integration and/or end-to-end tests.
- [x] I have informed others of any required `.env` changes files if required and changed the `.env-dist` accordingly.
- [x] I have included comments in the code to elaborate on what is not self-evident from the code itself, including references to issues and discussions online, or implicit behavior of an interface.

---

## Checklist for code reviewers:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_code.md) into your comment.

---

## Checklist for QA:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_qa.md) into your comment.

### CODE REVIEW

File: octopoes/octopoes/models/ooi/reports.py
Comment (ammar92) [inline octopoes/octopoes/models/ooi/reports.py:55]: I think `cron_expression` is a more appropriate name

**Review by ammar92 — APPROVED** [review]

File: octopoes/octopoes/models/ooi/reports.py
Comment (Rieven) [reply octopoes/octopoes/models/ooi/reports.py:55]: @underdarknl I think we need to remove this attribute from the recipe. It woudl be better if this lives in the scheduler. We can always pull all schedules and change the cron expressiom at the scheduler, otherwise f there is an update we need to update this variable in 2 places.

**Comment by stephanie0x00** [comment]
## Checklist for QA:

- [x] I have checked out this branch, and successfully ran a fresh `make reset`.
- [ ] I confirmed that there are no unintended functional regressions in this branch:
  - [x] I have managed to pass the onboarding flow
  - [x] Objects and Findings are created properly
  - [x] Tasks are created and completed properly
- [ ] I confirmed that the PR's advertised `feature` or `hotfix` works as intended.
- [ ] I checked the logs for errors and/or warnings and made issues where necessary

### What works:
The button "Add a reportRecipe" works. 

### What doesn't work:
This PR could use some love to make it easier to understand, as I haven't managed to actually enter a valid ReportRecipe. 

- Could you add helper functionalities to each of the fields on the ReportRecipe page? Each of the fields should contain a useful and easy to understand description of what the field is used for and what the user could input as a valid data. If necessary you can point to relevant documentation, but ideally that shouldn't be necessary at all. 
- Three of the field names are not 'human readable' but are direct input from the code. These are the recipe_id, input recipe and report_types. 

image

### Bug or feature?:
- The error messages are very developer-like when you add dummy data, as shown in the screenshot(s) below. 

image


 - recipe_id: I do not know what this means ` Error: Input should be a valid UUID, invalid length: expected length 32 for simple format, found 1 `. Does that mean it should be a uuid of 32 characters? Ifso, why don't you just generate one for me. A less technical user will not be able to generate a uuid very easily (if they even know what this is). If at all possible the preferred way would be not to bother the user with uuids anywhere in the UI, only with 'name' fields where they can give it a name of their choice. 


 - input_recipe: same as above. This is a technical error message that doesn't help the user to fix their problem. Give examples of formats that are acceptable. This can also be done through the helper functionality. Then the error message can simply state what doesn't match to the expected input. According to w3schools the input as shown below is a valid dictionary, but this appears not to be the case. 


image


 - report_types: again a very developer error message (step 1). This also is confusing. I next enter a valid list, now magically my input_recipe and recipe_id are not an issue anymore but I get an error that I should enter valid JSON (step 2). 
 
Step 1:
image

Step 2:
image

File: octopoes/octopoes/models/ooi/reports.py
Comment (noamblitz) [inline octopoes/octopoes/models/ooi/reports.py:52]: should include name with placeholders for date

Comment (Rieven) [reply octopoes/octopoes/models/ooi/reports.py:52]: The name belongs to the report OOI and not the recipe OOI as I can see that we can make 1 recipe for multiple reports with different names

Comment (noamblitz) [reply octopoes/octopoes/models/ooi/reports.py:52]: But you choose the name (with place holder) during the creation of the recipe, not after a run of creating the report.

Comment (Rieven) [reply octopoes/octopoes/models/ooi/reports.py:52]: The user choses the report name at the final step of the report creation, before it creates the report OOI and at that moment we create the recipe + report and also add the recipe to the report ooi as reference. Or do you mean something different?

Comment (Rieven) [reply octopoes/octopoes/models/ooi/reports.py:52]: For concatenated report, let's say we choose all report types, around 11 types. The final step shows 11 report names to change. we create 11 report oois with 1 recipe ooi, then we have 11 different report names that points to that 1 recipe.

File: octopoes/octopoes/models/ooi/reports.py
Comment (underdarknl) [reply octopoes/octopoes/models/ooi/reports.py]: as discussed, it stays in the ooi.

File: octopoes/octopoes/models/ooi/reports.py
Comment (underdarknl) [reply octopoes/octopoes/models/ooi/reports.py:52]: I would expect the recipe to have the name _including_ placeholders, for itself, and its child-reports.

Comment (noamblitz) [reply octopoes/octopoes/models/ooi/reports.py:52]: Jup agree, for the child reports, mostly placeholders should be used. Something like `{report_type}, for {input_ooi} on {date}` parent report could then be called something `My report on {date}`

Comment (Rieven) [reply octopoes/octopoes/models/ooi/reports.py:52]: Yes, but this is already the case. See https://github.com/minvws/nl-kat-coordination/blob/200eed578ca6f1675f4d027d47e4f2d974ba438a/octopoes/octopoes/models/ooi/reports.py#L26

**Comment by noamblitz** [comment]
Let's not merge this before complete development for report scheduling is done, its way easier to make changes to the OOI model before merging than after.

**Comment by Rieven** [comment]
> Let's not merge this before complete development for report scheduling is done, its way easier to make changes to the OOI model before merging than after.

The base branch is `feature/report-recipe`

**Comment by noamblitz** [comment]
> > Let's not merge this before complete development for report scheduling is done, its way easier to make changes to the OOI model before merging than after.
> 
> The base branch is `feature/report-recipe`

I missed this, sorry. Good idea :)

**Comment by noamblitz** [comment]
I think there should be a name for the sub-reports as well, not sure if users are able to choose the names of subreports.

File: octopoes/octopoes/models/ooi/reports.py
Comment (noamblitz) [inline octopoes/octopoes/models/ooi/reports.py:63]: I would say that the report_name_format should be part of the human readable. The id is not really readable.

File: octopoes/octopoes/models/ooi/reports.py
Comment (noamblitz) [inline octopoes/octopoes/models/ooi/reports.py:56]: Should maybe be optional? Not sure.

Comment (Rieven) [reply octopoes/octopoes/models/ooi/reports.py:56]: I think we must set a name for the report and subreports, otherwise the PDF headings would be incomplete and not visible. Also not clear in json export which report belongs to what.

Comment (noamblitz) [reply octopoes/octopoes/models/ooi/reports.py:56]: Okay

File: octopoes/octopoes/models/ooi/reports.py
Comment (Rieven) [reply octopoes/octopoes/models/ooi/reports.py]: Ok then we need the instance name otherwise it would not make sense.

File: octopoes/octopoes/models/ooi/reports.py
Comment (noamblitz) [inline octopoes/octopoes/models/ooi/reports.py:63]: for consistency please use `format_reference_human_readable()`

File: octopoes/octopoes/models/ooi/reports.py
Comment (noamblitz) [inline octopoes/octopoes/models/ooi/reports.py:56]: comment can be removed

**Pull request merged**
